### PR TITLE
Illegal character '&' parsing

### PR DESCRIPTION
**Pull request by masaruhoshi**
__Issue__
Garanti gateway seems to be returning XML responses with invalid characters (non-html entities).

__Solution__
Check if the raw response from Garanti is parseable. If it's not, try to clear it removing occurrences of '&' characters. Perhaps this may be happening due to a badly generated response exclusively from Garanti, but I decided placing the method `strip_invalid_xml_chars` in `Gateway` as it may be useful in other gateways.

__Review__
@girasquid @aprofeit cc: @Shopify/payments

### CODE REVIEW

File: lib/active_merchant/billing/gateway.rb [hunk 1]
```diff
@@ -212,6 +212,16 @@ def user_agent
         })
       end
 
+      def strip_invalid_xml_chars(xml)
+        begin
+          REXML::Document.new(xml)
+        rescue REXML::ParseException
+          xml = xml.gsub(/&(?!(?:[a-z]+|#[0-9]+|x[a-zA-Z0-9]+);)/, '&amp;')
```
Comment (girasquid) [inline lib/active_merchant/billing/gateway.rb]: If I'm following this correctly - if you have valid XML, you'll return a `REXML::Document` - but if you have invalid XML that raises a `REXML::ParseException`, you'll get a string back? Can you make the return values consistent?

This should also be going through Nokogiri; we'd like to use that over REXML going forward.

**Comment by girasquid** [comment]
Can you keep `strip_invalid_xml_chars` on the Garanti gateway for now? We can move it to `Gateway` when another gateway needs it, but for now we only know that Garanti does.

**Comment by masaruhoshi** [comment]
np! :ok_hand:

**Comment by aprofeit** [comment]
Can you link the original exception/issue? Just curious on what it looks like.

File: lib/active_merchant/billing/gateways/garanti.rb [hunk 2]
```diff
@@ -230,7 +230,7 @@ def commit(money,request)
       end
 
       def parse(body)
-        xml = REXML::Document.new(body)
+        xml = REXML::Document.new(strip_invalid_xml_chars(body))
```
Comment (j-mutter) [inline lib/active_merchant/billing/gateways/garanti.rb]: Why rebuild the REXML doc if it's being done in the method? Maybe the strip method should just run the gsub no matter what?

**Comment by masaruhoshi** [comment]
@aprofeit, they are [here] (https://errbit-virginia.cs.shopify.com/apps/54b810d6f9fec1380b000003/problems/553a3737f9fec10bf600005f) and [here] (https://errbit-virginia.cs.shopify.com/apps/54b810d6f9fec1380b000003/problems/555253a2f9fec163c400023a)

File: lib/active_merchant/billing/gateways/garanti.rb (hunk 2 above)
Comment (masaruhoshi) [inline lib/active_merchant/billing/gateways/garanti.rb]: That's actually a good point.

**Comment by masaruhoshi** [comment]
@girasquid , @aprofeit , @j-mutter , any more comments ?

File: lib/active_merchant/billing/gateways/garanti.rb [hunk 3]
```diff
@@ -251,6 +251,10 @@ def success?(response)
         response[:message] == "Approved"
       end
 
+      def strip_invalid_xml_chars(xml)
+        xml.gsub(/&(?!(?:[a-z]+|#[0-9]+|x[a-zA-Z0-9]+);)/, '&amp;')
```
Comment (girasquid) [inline lib/active_merchant/billing/gateways/garanti.rb]: This regex is a bit hard to parse - would the regex [here](http://stackoverflow.com/a/17237801/240418) do the same job but be more clear?

Comment (masaruhoshi) [inline lib/active_merchant/billing/gateways/garanti.rb]: That regex covers basically '&', '<', '>', double quote and apostrophe. Any other html entity (such as &atilde; &ccedil; etc) will have wrongfully converted their '&' with '&amp;'.

www.debuggex.com is a really option to understand complex regex.

Comment (girasquid) [inline lib/active_merchant/billing/gateways/garanti.rb]: Fair enough!

**Comment by girasquid** [comment]
:+1:

**Comment by aprofeit** [comment]
:+1:
